csrc/scheduler/normalization.cpp

```diff
@@ -943,200 +978,176 @@ std::shared_ptr<ReductionParams> outerPersistentHeuristic(
-      // Available unrolling based on size of data type
-      (int64_t)16 / (int64_t)max_input_dtype_size,
-      // Reduce unrolling if we have many inputs, start reduction at 4
-      // inputs
-      scheduler_utils::lastPow2(
-          scheduler_utils::safeDiv((int64_t)n_tensor_inputs, 4)));
-
-  // If we have more than a wave of blocks, put parallelism into unrolling
-  if (target_blocks > device_multiprocessor_count) {
-    target_unroll = std::min(
-        max_unroll, ceilDiv(target_blocks, device_multiprocessor_count));
-    target_blocks = ceilDiv(target_blocks, target_unroll);
-  }
-
-  // Cap target blocks to 4 waves
-  target_blocks = std::min(target_blocks, device_multiprocessor_count * 4);
-
-  if (target_blocks * target_unroll * max_threads_in_block < n_elems) {
-    // targetting 4 waves, so try to use a quarter of available threads
-    max_threads_in_block = std::min(
-        ceilDiv(n_elems, target_blocks * target_unroll),
-        ceilDiv(device_max_threads_per_multiprocessor, (int64_t)4));
-  }
-
-  // Round up to nearest warp.
-  if (max_threads_in_block % warp_size != 0) {
-    max_threads_in_block += warp_size - max_threads_in_block % warp_size;
-  }
-
   // Compute maximum number of reductions we could do in the same kernel based
   // on persistent buffer size
   const int64_t max_multi_reduction_factor = scheduler_utils::safeDiv(
       scheduler_utils::register_file_size, max_persistent_buffer_size);
 
-  // To get to target threads:
-  // Prioritize
-  // (1) x dim in iter domain
-  // (2) unrolling in iter domain
-  // (3) y in reduction domain
-  // To get target blocks:
-  // Prioritize
-  // (1) x dim in multiple outputs
-  // (2) y dim in multiple reductions - need to flip unrolling to reduction
-  // domain for this
-
-  // Threads for reduction
-  int64_t bdimy = 1;
-  // Threads for output
-  int64_t bdimx = 1;
-
-  // Unroll amount
-  int64_t inner_reduction_unroll_factor = 1;
-  int64_t iter_unroll_factor = 1;
-
-  // If we only use a warp, can we get iter domain unrolling?
-  bdimx = std::min(max_multi_reduction_factor, warp_size);
-  // Round down if it didn't hit a full warp
-  if (bdimx < warp_size) {
-    bdimx = scheduler_utils::lastPow2(bdimx);
+  struct HeuristicParams {
+    // Iteration dim, each CTA covers [bdimx] * [iter_unroll_factor] reductions.
+    // Needs total_iteration_numel / (bdimx * iter_unroll_factor) CTAs.
+    HeuristicParaWrapper iter_unroll_factor;
+    HeuristicParaWrapper bdimx;
+    // Reduction dim, each thread do [batches_per_block * redu_unroll_factor]
+    // serial reductions, then do block reductions along [bdimy].
+    // Total_reduction_numel <= bdimy [dynamic] * batches_per_block *
+    // redu_unroll_factor
+    HeuristicParaWrapper redu_unroll_factor;
+    HeuristicParaWrapper batches_per_block;
+    HeuristicParaWrapper bdimy;
+    void verify() {
+      TORCH_INTERNAL_ASSERT(
+          !iter_unroll_factor.isMutable(), "iter_unroll_factor is not set.");
```

Comment:
Being set and mutable don't seem to conflict. I suggest the error message should be "not finalized".

Reply:
changed.